realm/realm-library/src/main/java/io/realm/internal/objectstore/OsSubscriptionSet.java

```
@@ -0,0 +1,236 @@
+                    waitForSynchronization(timeOut, unit);
+                    callback.onStateChange(OsSubscriptionSet.this);
+                } catch (Exception e) {
+                    callback.onError(e);
+                }
+            }
+        });
+        return new RealmAsyncTaskImpl(future, stateListenerExecutor);
+    }
+
+    @Override
+    public SubscriptionSet update(UpdateCallback action) {
+        OsMutableSubscriptionSet mutableSubs = new OsMutableSubscriptionSet(
+                nativeCreateMutableSubscriptionSet(nativePtr),
+                schema,
+                stateListenerExecutor,
+                updateExecutor
+        );
+        action.update(mutableSubs);
+        long newSubscriptionsSet = mutableSubs.commit();
+        // Once commit succeed, replace the current SubscriptionSet pointer with
+        // the new one. If the commit fails, the MutableSubscriptionSet will be
+        // GC'ed and released and the SubscriptionSet will keep the state before
+        // the UpdateCallback was called.
+        long oldPointer = nativePtr;
+        nativePtr = newSubscriptionsSet;
+        nativeRelease(oldPointer);
+        return this;
+    }
+
+    @Override
+    public RealmAsyncTask updateAsync(UpdateAsyncCallback callback) {
+        Future<?> future = updateExecutor.submit(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    SubscriptionSet updatedSubscriptions = update(callback);
+                    callback.onSuccess(updatedSubscriptions);
+                } catch (Exception exception) {
+                    callback.onError(exception);
+                }
+            }
+        });
+        return new RealmAsyncTaskImpl(future, updateExecutor);
+    }
+
+    @Override
+    public void refresh() {
+        nativeRefresh(nativePtr);
+    }
+
+    @Override
+    public Iterator<Subscription> iterator() {
+        // This iterator does not support removals.
```

We could implement the `remove` method to throw a `UnsupportedOperationException` exception, but I guess nobody is going to mess with this iterator.